### PR TITLE
Fix console, Math matching

### LINKED ISSUE #354: Console keyword causing problems
I came here from atom/language-html.

Basically the `console` keyword used within `<script>` tags is causing problems with syntax highlighting.

error

The problem is due to using the keyword console directly.

See below, when I replace console with konsole, it works:

working

Link to lanugage-html issue: https://github.com/atom/language-html/issues/111

### PR DESCRIPTION
**Pull request by MaximSokolov**
Fixes #354
Maybe @kartikanand knowns a better solution

Refs #355
Refs #344 (wrong scopes with `$`)

### CODE REVIEW

**Comment by kartikanand** [comment]
Yours is a way better and clean solution. One thing I don't understand, why the need for `(?!\\$)` in `begin` ?

**Comment by MaximSokolov** [comment]
Because of incorrect word boundaries(issue #344 ):
```js
console$
```
However `(?!\\$)'` doesn't work for `Math` for some reason
```coffee
'begin': '\\bMath\\b(?!\\$)'
```
```js
Math$
```

**Comment by CentricStorm** [comment]
Still more problems to fix:

console-math-problems

File: grammars/javascript.cson [hunk 1]
```diff
@@ -702,11 +702,11 @@
   {
     # console
     # console.log(arg1, "arg2", [...])
-    'begin': '\\bconsole\\b'
+    'begin': '\\bconsole\\b(?!\\$)'
```
Comment (kartikanand) [inline grammars/javascript.cson]: Maybe we can do something like -

`'begin': '(?<!\\$)\\bconsole\\b(?!\\$|//|/\*)'`

File: grammars/javascript.cson [hunk 2]
```diff
@@ -702,11 +702,11 @@
...
     'beginCaptures':
       '0':
         'name': 'entity.name.type.object.console.js'
-    'end': '(?!\\G)'
+    'end': '(?=(?!\\s*\\.)\\S)'
     'patterns': [
       {
         'begin': '\\s*(\\.)\\s*(assert|clear|debug|error|info|log|profile|profileEnd|time|timeEnd|warn)\\s*(?=\\()'
```
Comment (kartikanand) [inline grammars/javascript.cson]: Here we can capture all properties?

`'begin': '\\s*(\\.)\\s*(\w+)\\s*(?=\\()'`

Comment (kartikanand) [inline grammars/javascript.cson]: Here we can capture all properties?

'begin': '\\s*(\\.)\\s*(\w+)\\s*(?=\\()'

Comment (kartikanand) [inline grammars/javascript.cson]: Here we can capture all properties?

`'begin': '\s(\.)\s(\w+)\s*(?=\()'`

**Comment by MaximSokolov** [comment]
> But for some reason it's not working for new lines, maybe begin doesn't take newlines into account: 
```js
console
: 'asd'
```

The tokenizer works line by line, so nothing you can do about this issue

> Should not be highlighted: 
```js
console//
console/**/
```

Looks legit to me
image

File: grammars/javascript.cson [hunk 3]
```diff
@@ -702,14 +702,21 @@
...
+        (?! \\s*(\\.)\\s*
+          (assert|clear|debug|error|info|log|profile|profileEnd|time|timeEnd|warn)
+          \\s*\\(
+        ) \\S
+      )
+    '''
     'patterns': [
       {
-        'begin': '\\s*(\\.)\\s*(assert|clear|debug|error|info|log|profile|profileEnd|time|timeEnd|warn)\\s*(?=\\()'
+        'begin': '\\s*(\\.)\\s*(\\w+)\\s*(?=\\()'
```
Comment (50Wliu) [inline grammars/javascript.cson]: Why was this changed?

Comment (MaximSokolov) [inline grammars/javascript.cson]: `\w` is enough here, as I modified `end` pattern to allow regular functions/properties.

**Comment by CentricStorm** [comment]
> Looks legit to me


Yeah I should have been clearer - the comments in that example image were only referring to the lines immediately after them - `console//` is working fine :thumbsup:.

Unfortunately there is still a problem though:

untitled

**Comment by MaximSokolov** [comment]
>
```js
console . loga()
```

Fixed, thanks!

**Comment by CentricStorm** [comment]
More issues:

untitled

**Comment by 50Wliu** [comment]
Looks good to me except for one thing.  Adding a comment before the period does not cause the comment to get recognized, eg `console/*hi*/.log()`.  This was evidently working according to @CentricStorm's previous screenshot, but now it isn't.

Some things that we might want to address in future PRs though:
* `console` and `Math`-specific methods aren't being tokenized with `entity.name.function.js`.
* Adding a comment after the period causes the match to end prematurely, eg `console./*hi*/log()`.

**Comment by CentricStorm** [comment]
> Looks good to me except for one thing. Adding a comment before the period does not cause the comment to get recognized, eg `console/*hi*/.log()`. This was evidently working according to @CentricStorm's previous screenshot, but now it isn't.

Working fine for me on Atom 1.8.0:

untitled

> - `console` and `Math`-specific methods aren't being tokenized with `entity.name.function.js`.

What should they be scoped as? With the `entity.name.function.js` scope, they will be indistinguishable from normal function calls. Some colour schemes apply styling to `.entity.name.function`, which is higher precedence than `.support.function`, so will override.

> - Adding a comment after the period causes the match to end prematurely, eg `console./*hi*/log()`.

Some things are impractical to do with regular expressions. If 100% syntax highlighting accuracy is desired then it would be easier to change to a language-specific parser.

**Comment by MaximSokolov** [comment]
> Adding a comment before the period does not cause the comment to get recognized, eg
```js
console/*hi*/.log()
```

Are you sure you've tested this branch? It should be working in both f625823 and 7bddd57
Now `.log()` is also scoped properly (`support.function`)

**Comment by MaximSokolov** [comment]
Agree with @CentricStorm about scopes and `console./*hi*/log()` case

**Comment by 50Wliu** [comment]
I'm pretty sure I was...I specifically checked out the branch on Github Desktop :confused:.  I'll give it another go today.
